anndata/core/anndata.py

```diff
@@ -1829,6 +1829,30 @@ def concatenate(
 
         var = pd.DataFrame(index=var_names)
 
+        if join == 'inner':
```

Comment:
@Koncopd, great! But does it really have to be that long? Can you just use `.drop_replicates` (`adata.var.drop_duplicates`) on the joint object and simplify this heavily? I would have expected a 2 line solution with no indendation... :wink:

This is way too much code and way too complicated. 🙂

Reply:
I can do as you say, but i thought we need not just remove duplicates, but remove duplicates and leave the original column name (without "-0") only if  **all** anndatas have the same var column.
This is the reason i wrote it in this way. 
If we don't need it, i willjust add (however it will require to copy the whole var i believe)
```
var = var.T
var.drop_duplicates(inplace=True)
var = var.T
```

Reply:
Copying the whole var is not a problem. Minimizing code complexity is more important, here. And yes, we need to leave the original column name (without "-0"). :)

Reply:
Thank you!

Reply:
@falexwolf 
It is still unclear to me what are the right conditions for removing "duplicated" columns.
What if columns have different names but the same values? drop_duplicates drops such columns, but is it desired?
For example,

```
    adata1 = AnnData(X1,
                     {'obs_names': ['s1', 's2'],
                      'anno1': ['c1', 'c2']},
                     {'var_names': ['a', 'b', 'c'],
                      'annoA': [0, 1, 2]},
                     layers={'Xs': X1})
    adata2 = AnnData(X2,
                     {'obs_names': ['s3', 's4'],
                      'anno1': ['c3', 'c4']},
                     {'var_names': ['d', 'c', 'b'],
                      'annoA': [0, 1, 2]},
                     layers={'Xs': X2})
    adata3 = AnnData(X3,
                     {'obs_names': ['s1', 's2'],
                      'anno2': ['d3', 'd4']},
                     {'var_names': ['d', 'c', 'b'],
                      'annoB': [0, 1, 2]},
                     layers={'Xs': X3})

    # inner join
    adata = adata1.concatenate(adata2, adata3)
```
The result for `adata.var_keys()` will be `['annoA-0', 'annoA-1']`, it removes `'annoB'`.
Is this approach correct?

Also, it is not clear when exactly we should leave the original column name, and where we shouldn't.

Reply:
Yeah, we should rather fix the code that creates the duplicated columns.

When concatenating and there’s identical columns with identical names, they should be merged, I think.

Why do we copy columns manually instead of using `var1.join(var2)`?

Reply:
Yes, thanks, this seems like the right approach. 

Still unsure about naming of duplicates and non-duplicates as there are batch categories and so on in the function. If we need to leave the original name, then what to do with columns that have the same name but different content? Like the former will have the original name, the latter will have "name+number of the batch". This is not systematic.

Reply:
I think renaming them all and throwing a warning would be a good approach.

Reply:
It's really just about the rows of `.var` that have exactly the same column names and exactly the same values. This is a big use case. You could even simplify to making this reduction only if all rows in `.var` meet this condition.

If not, it makes sense to be as verbose as we are in the current case. All intermediate cases add complexity.

Reply:
@falexwolf , 

> rows of .var that have exactly the same column names and exactly the same values

> all rows in .var meet this condition

Do you mean columns instead of rows?

Reply:
Yes, sorry, I mean columns of the `var` DataFrame (when saying "rows", I had the whole AnnData object in mind). `gene_ids-0`, `gene_ids-1`, etc. doesn't make sense if they store exactly the same content. This is the main use case.